### PR TITLE
Ruby 3.3 (and dependents)

### PR DESCRIPTION
**Pull request by faisal**
Update the Ruby formula to build the just-released [Ruby 3.3](https://www.ruby-lang.org/en/news/2023/12/25/ruby-3-3-0-released/).

- Updated Ruby to 3.3, including updating RubyGems to 3.5.3
- Removed bison and readline dependencies since Ruby [no longer uses them](https://bugs.ruby-lang.org/issues/19637)
- Bumped revision number for formulae that `depends_on "ruby"` and require an updated libruby version.
- Renamed ruby@3.2 Alias (that pointed to ruby formula) to ruby@3.3, and create new Ruby@3.2 formula for legacy support

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

**Comment by SMillerDev** [comment]
> Bumped revision number for anything that depends_on "ruby", so as to trigger rebuild

Can you revert this? Not everything has to be rebuilt, only things that link to ruby libraries might need a rebuilt.

**Comment by faisal** [comment]
> > Bumped revision number for anything that depends_on "ruby", so as to trigger rebuild
> 
> Can you revert this? Not everything has to be rebuilt, only things that link to ruby libraries might need a rebuilt.

My latest push is just Ruby and its Aliases, backing out (squashed) the bumps to all the other formula. Unfortunately it looks like in doing so I may have broken a fix @crunchtime-ali added for k9s right before I pushed the squashed commit.

**Comment by faisal** [comment]
> My latest push is just Ruby and its Aliases, backing out (squashed) the bumps to all the other formula. Unfortunately it looks like in doing so I may have broken a fix @crunchtime-ali added for k9s right before I pushed the squashed commit.

This k9s break looks like merge noise, which came out of the PR when I rebased to clean it up and pick up https://github.com/Homebrew/homebrew-core/pull/158215/commits/981c7ec1567cbd2347c8fb4b7da495356cfbdcb9.

**Comment by SMillerDev** [comment]
Why bump the revisions again? I really doubt that would be needed for most of the formulae in the list

**Comment by faisal** [comment]
> Why bump the revisions again? I really doubt that would be needed for most of the formulae in the list

Most of the formulae that `depends_on "ruby"` actively depend on the specific version of Ruby that was extant at build time, which shows up as a large number of linkage failures in CI. The revision bumps update that version expectation to Ruby 3.3.

I'd initially done a blanket rebuild in the initial PR. The new bumps are more considered, and only for formulae that failed to build in CI _because of the version mismatch_:

- For all but _weechat_, the dependent formulae link to the version of `libruby` available in `/opt/homebrew/opt/ruby/lib/` at build time: formerly `libruby.3.2` but now  `libruby.3.3` (`.dylib` on macOS). You can see these failures in the unfortunate number checks I triggered to get here, e.g. https://github.com/Homebrew/homebrew-core/actions/runs/7322797920/job/19944994270#step:3:1341

- _wechat_ declares a list of Ruby versions to work with, so for that formula I added a line to [add Ruby 3.3 to the list](https://github.com/Homebrew/homebrew-core/pull/158215/files#diff-5c907be2daed29486151f39662e38a711da8237b122882f0f19901ba28af2bab) if the list doesn't yet include Ruby 3.3.

**Comment by faisal** [comment]
> Most of the formulae that `depends_on "ruby"` actively depend on the specific version of Ruby that was extant at build time, which shows up as a large number of linkage failures in CI. The revision bumps update that version expectation to Ruby 3.3.

We had a similar need in [Ruby 3.2](https://github.com/Homebrew/homebrew-core/pull/118926/commits), and [Ruby 3.1](https://github.com/Homebrew/homebrew-core/pull/92024/commits). I expect we also had them in prior major version bumps but I haven't tracked down all the PRs.

**Comment by faisal** [comment]
I'm somewhat at a loss on these remaining failing checks. For solargraph I could explicitly declare an xz dependency, but that's purely speculative since it should work as-is.

**Comment by faisal** [comment]
> * _wechat_ declares a list of Ruby versions to work with, so for that formula I added a line to [add Ruby 3.3 to the list](https://github.com/Homebrew/homebrew-core/pull/158215/files#diff-5c907be2daed29486151f39662e38a711da8237b122882f0f19901ba28af2bab) if the list doesn't yet include Ruby 3.3.

FWIW I also put up https://github.com/weechat/weechat/pull/2055 to update Weechat's own build system to detect the new Ruby version. If they take it then this patch here should still work: we'll still need the rebuild, and the `install` version check replacement should noop.

**Comment by faisal** [comment]
> I'm somewhat at a loss on these remaining failing checks. For solargraph I could explicitly declare an xz dependency, but that's purely speculative since it should work as-is.

Explicitly declaring the xz dependency got rid of that particular failure.

**Comment by faisal** [comment]
At this point, I think this PR could use attention from someone familiar with the test-bot workflow, to parse why the x86 macOS and the conclusion checks are failing.

**Comment by faisal** [comment]
It looks like the Intel tests are reliably timing out at around 1:30. Given that the remainder seems to be passing, would it make sense to land the PR as-is, and then fix any breakage we find afterwards? Or to remove the dependents from the PR, land the newRuby, and then follow up with PRs to fix dependents?

**Comment by faisal** [comment]
I created https://github.com/Homebrew/homebrew-core/pull/158480 to be able to reason about what happens if we just patch Ruby and not track down all the dependencies. We can close it or this one, as appopriate.

**Comment by ryenus** [comment]
> I created #158480 to be able to reason about what happens if we just patch Ruby and not track down all the dependencies. We can close it or this one, as appopriate.

Not to judge, but I for one think it's reasonable, especially in case of minor issues only. Better than what's going on in the Python camp :-)

**Comment by SMillerDev** [comment]
> It looks like the Intel tests are reliably timing out at around 1:30.

Yes, this is by design. It needs a label to resolve that

**Comment by faisal** [comment]
> > I created #158480 to be able to reason about what happens if we just patch Ruby and not track down all the dependencies. We can close it or this one, as appopriate.
> 
> Not really to judge, but I for one think it's reasonable, especially in case of minor issues only. Better than what's going on in the Python camp :-)

I think it's worth the discussion.

I figure if that one gets approved then I'll turn around and fire off PRs to resolve every dependent I know of. Some of them are fairly trivial and some require patches, but loading them all in this one PR was glacial before it was impossible.

**Comment by faisal** [comment]
> > It looks like the Intel tests are reliably timing out at around 1:30.
> 
> Yes, this is by design. It needs a label to resolve that

How do we get it labeled?

**Comment by SMillerDev** [comment]
We need to wait until a space in the longer running queue is available.

**Comment by faisal** [comment]
> Maintainers won't merge a PR if dependents tests didn't pass, unless explicitly marked (labeled) for this, for good reasons chosen by them. Opening yet another PR where dependencies testing will obviously break as fixes present here are missing is only adding noise and loading the CI for nothing.

OK, I've closed that one. In hindsight, it seems this would be better handled with a staging branch. Would it make sense to rename the branch behind this PR now, and 

> I also suggested you to drop the head block that's not needed anymore in a pending review above, but you doesn't seem to consider it.

I missed that comment and don't see one now (possibly having inadvertently overwritten with a subsequent force push). To which formula were you referring?

**Comment by faisal** [comment]
> > Maintainers won't merge a PR if dependents tests didn't pass, unless explicitly marked (labeled) for this, for good reasons chosen by them. Opening yet another PR where dependencies testing will obviously break as fixes present here are missing is only adding noise and loading the CI for nothing.
> 
> OK, I've closed that one. In hindsight, it seems this would be better handled with a staging branch.

Would it make sense for me to rename this branch to something like ruby-3-3-staging, and proceed with new PRs using that model?

**Comment by SMillerDev** [comment]
> Would it make sense for me to rename this branch to something like ruby-3-3-staging, and proceed with new PRs using that model?

No, since staging branches can not be done by normal contributors. They require special permissions.
Nothing here requires a staging branch though, just some patience to wait for the next long build slot.

**Comment by faisal** [comment]
> You just have to look at the pending reviews, visible from GitHub. That head block is still around in the ruby formula.

I don't see any pending reviews. The only place the string "pending" appears is in this comment thread. I don't know if this discrepancy is due to our respective roles in the PR or something else.

Was your suggestion to remove the head block entirely, to replace it with a single line declaration, or something else?

Were there other suggestions?

**Comment by bevanjkay** [comment]
@Porkepix if your review comment shows as 'pending' (with a yellow label) that means that it is only visible to you and needs to be submitted still. You either write the comment as a single comment or if attaching it to a full review you then have to submit the full review.

File: Formula/r/ruby.rb
Comment (faisal) [reply Formula/r/ruby.rb:38]: line 6?

Comment (faisal) [reply Formula/r/ruby.rb:38]: Line redacted, and revision bumps reasoning added in commits that I'll have to force push shortly.

**Comment by SMillerDev** [comment]
Please stop adding merge commits, since those trip up CI.

**Comment by p-linnane** [comment]
@faisal Thanks for your patience. I've started CI with the long build label.

**Comment by faisal** [comment]
> @faisal Thanks for your patience. I've started CI with the long build label.

Thanks! If we find things issues in this CI and need to iterate, what's the best way to do so?

**Comment by p-linnane** [comment]
If changes are needed we can push them here and rerun CI. I've added a label which will make CI continue even if a build fails so we can capture all errors if there are any.

**Comment by p-linnane** [comment]
`sugarjar` will need a revision bump. Not sure what is going on with `licensefinder`. It appears to be using system Ruby despite having a dependency on our `ruby` formula. Any ideas @Homebrew/core?

**Comment by faisal** [comment]
> `sugarjar` will need a revision bump. Not sure what is going on with `licensefinder`. It appears to be using system Ruby despite having a dependency on our `ruby` formula. Any ideas @Homebrew/core?

We may need someone from @pivotal to look at LicenseFinder.

**Review by MikeMcQuaid — APPROVED** [review]
This looks good to me, nice work @faisal! I'd suggest we make still failing formulae `depends_on "ruby@3.2"` for now instead to get this merged.

**Comment by faisal** [comment]
After cherry-picking @branchvincent's fix, I still get:

```
$ brew test licensefinder
==> Testing licensefinder
==> bundle install
==> /opt/homebrew/Cellar/licensefinder/7.1.0_1/bin/license_finder
<internal:/opt/homebrew/Cellar/ruby/3.3.0/lib/ruby/3.3.0/rubygems/core_ext/kernel_require.rb>:127:in `require': cannot load such file -- racc/parser.rb (LoadError)
```

For now I've tagged this one `depends_on "ruby@3.2"`, and will continue to poke at it until I see what else CI finds (or we get a clean run, 🤞).

**Comment by faisal** [comment]
I found a fix for licensefinder's remaining test issue. I'm holding on the results of the current run before pushing. When I do, should I push it as an additional commit, or rewrite that last final licensefinder commit and force push?

**Comment by faisal** [comment]
The licensefinder, inko, and mailcatcher commits should address all the breaks from the most recent CI run. I did not explicitly merge master, and am not sure what action I took that caused such a merge, but I'm leaving it alone rather than cause any further churn on CI.

**Comment by faisal** [comment]
> The licensefinder, inko, and mailcatcher commits should address all the breaks from the most recent CI run. I did not explicitly merge master, and am not sure what action I took that caused such a merge, but I'm leaving it alone rather than cause any further churn on CI.

From here would it be better to a) proceed with what I have, or b) to rebase on master and then squash the licensefinder, mailcatcher, and inko commits (plus one more fix for inko)?

**Comment by p-linnane** [comment]
> I found a fix for licensefinder's remaining test issue. I'm holding on the results of the current run before pushing. When I do, should I push it as an additional commit, or rewrite that last final licensefinder commit and force push?

Have 1 commit per formula please.

**Comment by faisal** [comment]
> > I found a fix for licensefinder's remaining test issue. I'm holding on the results of the current run before pushing. When I do, should I push it as an additional commit, or rewrite that last final licensefinder commit and force push?
> 
> Have 1 commit per formula please.

Thanks. After the current CI run finishes I will squash commits so there's a 1:1 relationship.

**Comment by faisal** [comment]
https://github.com/Homebrew/homebrew-core/actions/runs/7406867486/job/20156199574?pr=158215

**Comment by faisal** [comment]
It looks like setup_dup_runners is stalled. Can it be restarted without a new push?

**Comment by faisal** [comment]
I'm working on putting up a representative x86 Linux box, but please do chime in if you know of simple fixes for:

```
==> FAILED
Full install --ignore-dependencies --skip-post-install /ruby--3.3.0.x86_64_linux.bottle.tar.gz output
  Warning: `--ignore-dependencies` is an unsupported Homebrew developer option!
  Adjust your PATH to put any preferred versions of applications earlier in the
  PATH rather than using this unsupported option!
  
  ==> Fetching ruby
  ==> Pouring ruby--3.3.0.x86_64_linux.bottle.tar.gz
  ==> Skipping 'post_install' on request
  You can run it manually using:
    brew postinstall ruby
  Error: Failure while executing; `/home/linuxbrew/.linuxbrew/opt/ruby/bin/ruby -e print\ Gem.ruby_api_version` exited with 127. Here's the output:
```

...

```
Full install --ignore-dependencies --skip-post-install /ruby@3.2--3.2.2.x86_64_linux.bottle.tar.gz output
  Warning: `--ignore-dependencies` is an unsupported Homebrew developer option!
  Adjust your PATH to put any preferred versions of applications earlier in the
  PATH rather than using this unsupported option!
  
  ==> Fetching ruby@3.2
  ==> Pouring ruby@3.2--3.2.2.x86_64_linux.bottle.tar.gz
  ==> Skipping 'post_install' on request
  You can run it manually using:
    brew postinstall ruby@3.2
 ```

**Comment by faisal** [comment]
At this point this will need a rebase or merge to integrate updates that have landed recently.

**Comment by faisal** [comment]
Could I ask for help from someone with knowledge of how Linux Homebrew works, and in particular how its CI works?

I can't reproduce the current failure, and a 127 exit code seems odd since I didn't touch the logic for laying down the executable. Ruby@3.2 _should_ work exactly the same way as it ruby did when it was for version 3.2, and if anything changed I'd expect they changed for macOS as well.

**Comment by faisal** [comment]
FWIW, https://github.com/faisal/homebrew-core/tree/ruby_3_3_updated is a copy of the PR but rebased on master. I don't want to push to this PR and load up the CI until I have some confidence the Linux test issue is resolved (or one of the maintainers asks me to push).

**Comment by chenrui333** [comment]
kicked off a long build run.

**Comment by p-linnane** [comment]
The job to set up dependent testing has been hanging for hours despite multiple reruns. Not sure how to handle this.

**Comment by faisal** [comment]
> kicked off a long build run.

Looks like it canceled out at setup_dep_runners. Will someone restart it?

**Comment by p-linnane** [comment]
> Looks like it canceled out at setup_dep_runners. Will someone restart it?

We've tried a few times so far. It hangs for hours, when it should finish within a minute or two. We're trying to figure out what's going on.

**Comment by faisal** [comment]
While we're waiting, `rubygems` updated to 3.5.4 late last week. _If_ we have reason to re-push this PR, I'll include the newest version instead of 3.5.3. I don't believe the minor update is worth starting over, but please do let me know if you think otherwise.

**Comment by MikeMcQuaid** [comment]
@faisal So sorry for the delay here. Hold off any more pushes; we're working on resolving some CI issues.

**Comment by MikeMcQuaid** [comment]
I've rebased this on `master` (including @bo98's https://github.com/Homebrew/homebrew-core/pull/159398), made the `api_version` change needed for `ruby@3.2` and updated RubyGems to 3.5.4.

**Review by MikeMcQuaid — APPROVED** [review]

File: Formula/r/ruby@3.2.rb
Comment (Bo98) [inline Formula/r/ruby@3.2.rb:215]:
```suggestion
```

(but I can do this in a follow-up PR as it can be a syntax-only change - just noting here in case I forget)

**Review by p-linnane — APPROVED** [review]

**Comment by p-linnane** [comment]
Thank you for all the hard work here @faisal!

**Comment by faisal** [comment]
> Thank you for all the hard work here @faisal!

Thank you all for shepherding this through!

**Comment by MikeMcQuaid** [comment]
Thanks again @faisal, great work here!

**Pull request merged**
